### PR TITLE
Corrections after reworking backup/restore synchronization #2

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Follow-up to https://github.com/ClickHouse/ClickHouse/pull/70027 and https://github.com/ClickHouse/ClickHouse/pull/71715

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 380aea0fc3c6234a6cb1fa334c1fa1f5eef8fe23 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71912/380aea0fc3c6234a6cb1fa334c1fa1f5eef8fe23/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/acd4f46d9c703b5f8bb757555e6d085c0ea0a5fa/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by vitlibar** [comment]
Some tests are still failing ([see](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0ZXN0X25hbWUsIGNoZWNrX3N0YXJ0X3RpbWUsIHB1bGxfcmVxdWVzdF9udW1iZXIsIHJlcG9ydF91cmwKZnJvbSBjaGVja3Mgd2hlcmUgJzIwMjQtMTEtMTQnIDw9IGNoZWNrX3N0YXJ0X3RpbWUgYW5kIHRlc3RfbmFtZSBsaWtlICcldGVzdF9iYWNrdXBfcmVzdG9yZV9vbl9jbHVzdGVyJScgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIG9yZGVyIGJ5IGNoZWNrX3N0YXJ0X3RpbWU=)), so this PR will help them.

File: src/Backups/BackupCoordinationStageSync.cpp [hunk 1]
```diff
@@ -496,6 +499,9 @@ void BackupCoordinationStageSync::watchingThread()
             tryLogCurrentException(log, "Caught exception while watching");
         }
 
+        if (should_stop())
+            return;
```
Comment (vitlibar) [inline src/Backups/BackupCoordinationStageSync.cpp:503]: This is to help a cancelled backup to stop working faster, that must help `test_backup_restore_on_cluster/test_cancel_backup.py::test_cancel_backup`.

File: src/Backups/BackupCoordinationStageSync.cpp [hunk 2]
```diff
@@ -121,8 +121,7 @@ BackupCoordinationStageSync::BackupCoordinationStageSync(
 
     try
     {
-        concurrency_check.emplace(is_restore, /* on_cluster = */ true, zookeeper_path, allow_concurrency, concurrency_counters_);
-        createStartAndAliveNodes();
+        createStartAndAliveNodesAndCheckConcurrency(concurrency_counters_);
```
Comment (vitlibar) [inline src/Backups/BackupCoordinationStageSync.cpp:124]: The concurrency check was improved a bit, that must help `test_backup_restore_on_cluster/test_disallow_concurrency.py::test_concurrent_backups_on_different_nodes` to stop being flaky.

**Review by thevar1able — APPROVED** [review]

**Comment by vitlibar** [comment]
CI failures are not related:
- [test_storage_mongodb/test.py::test_secure_connection_uri](https://s3.amazonaws.com/clickhouse-test-reports/71912/380aea0fc3c6234a6cb1fa334c1fa1f5eef8fe23/integration_tests__asan__old_analyzer__[3_6].html) - already fixed by https://github.com/ClickHouse/ClickHouse/pull/71924
- [test_concurrent_threads_soft_limit/test.py::test_concurrent_threads_soft_limit_defined_50](https://s3.amazonaws.com/clickhouse-test-reports/71912/380aea0fc3c6234a6cb1fa334c1fa1f5eef8fe23/integration_tests__tsan__[6_6].html) - see https://github.com/ClickHouse/ClickHouse/issues/71672
- [03247_create-same-table-concurrently-with-atomic-engine](https://s3.amazonaws.com/clickhouse-test-reports/71912/380aea0fc3c6234a6cb1fa334c1fa1f5eef8fe23/stateless_tests__debug__[2_2].html)

**Pull request merged**